lib/Synchronise/Network/Server.hs

```diff
@@ -212,9 +230,137 @@ notify (RequestChange note) = do
+--   These workers cannot die, they simply log any errors and keep going.
+--
+worker :: Store store => store -> Configuration -> IO ()
+worker store cfg = go
+  where go = do
+          (_, item) <- liftIO $ getWorkBackoff store
+          case item of
+            WorkNotify fk -> do
+              liftIO . debugM logName $ "Processing a notifcation: " <> show fk
+              processNotification store cfg fk
+            WorkApplyPatch did diff -> do
+              liftIO . debugM logName $ "Processing a diff: " <> show did
+              processDiff did diff
+          go
+
+-- notifications
+
+processNotification :: Store store => store -> Configuration -> ForeignKey -> IO ()
+processNotification store cfg fk@(ForeignKey{..}) = do
+  let ds = do es <- M.lookup fkEntity (configEntities cfg)
+              _  <- M.lookup fkSource (entitySources es)
+              M.lookup fkSource (entitySources es)
+  case ds of
+    Nothing         -> liftIO . errorM logName $ "Unknown key in workqueue: " <> show fk
+    Just datasource -> do
+      ik    <- liftIO     $ lookupInternalKey store fk
+      doc   <- runDSMonad $ DS.readDocument datasource fk
+      let dss = L.delete datasource $ allDataSources cfg
+
+      liftIO $ case (ik, doc) of
+       (Nothing, Left  _) -> notifyProblem fk (SynchroniseUnknown "Unknown key. No Document")
+       (Nothing, Right d) -> notifyCreate  store dss fk d
+       (Just i,  Left  _) -> notifyDelete  store dss i
+       (Just i,  Right _) -> notifyUpdate  i
+
+-- | Creates a new internal document to reflect a new foreign change. Update
+--   all given data sources of the change.
+--
+--   Caller is responsible for: ensuring the datasources exclude the one from
+--   which the event originates.
+--
+notifyCreate
+  :: Store store => store
+  -> [DataSource] -- ^ Create in each of these data sources.
+  -> ForeignKey   -- ^ Using this foreign key.
+  -> Document     -- ^ And this document.
+  -> IO ()
+notifyCreate store datasources fk@(ForeignKey{..}) doc = do
+  infoM logName $ "CREATE: " <> show fk
+
+  -- Create an internal key associated with the new document
+  ik  <- createInternalKey store fkEntity
+  recordForeignKey store ik fk
+
+  -- Create an initial document
+  recordInitialDocument store ik doc
+
+  -- Update the document for all known entities
+  forM_ datasources updateDoc
+
+  where updateDoc ds = do
+          x  <- runDSMonad $ DS.updateDocument ds fk doc
+          case x of Left  e -> errorM logName (show e)
+                    Right _ -> return ()
+
+-- | Deletes internal document to reflect the foreign change. Update
+--   all given data sources of the change.
+--
+--   Caller is responsible for: ensuring the datasources exclude the one from
+--   which the event originates.
+--
+notifyDelete
+  :: Store store => store
+  -> [DataSource]
+  -> InternalKey
+  -> IO ()
+notifyDelete store datasources ik = do
+  infoM logName $ "DELETE: " <> show ik
+  forM_ datasources deleteDoc
+
```

Comment:
I think we also need a `deleteInternalKey`, `deleteInitialDocument`, etc. in here to get rid of the stale data?